worker/raft/raftclusterer/worker_test.go

```diff
@@ -84,15 +84,21 @@ var _ = gc.Suite(&WorkerSuite{})
 
 func (s *WorkerSuite) SetUpTest(c *gc.C) {
 	s.workerFixture.SetUpTest(c)
-
 	s.reqs = make(chan apiserver.DetailsRequest, 10)
-	s.hub.Subscribe(
+
+	// Use a local variable to send to the channel in the callback, so
+	// we don't get races when a subsequent test overwrites s.reqs
+	// with a new channel.
+	reqs := s.reqs
+	unsubscribe, err := s.hub.Subscribe(
 		apiserver.DetailsRequestTopic,
 		func(topic string, req apiserver.DetailsRequest, err error) {
 			c.Check(err, jc.ErrorIsNil)
-			s.reqs <- req
+			reqs <- req
 		},
 	)
+	c.Assert(err, jc.ErrorIsNil)
+	s.AddCleanup(func(c *gc.C) { unsubscribe() })
```

Comment:
The unsubscribe should be sufficient, as once unsubscribe occurs, any pending calls that haven't been processed get cleared out. So, once unsubscribe returns there will be no more calls to the callback. You don't need the local variable.

Reply:
Nope - that was what I thought initially too but it still failed. I think the problem the race checker is pointing out is:
* goroutine A (the main one) writes s.reqs (setting it to a new channel).
* goroutine B (inside the hub) reads s.reqs (to send on it)
* goroutine A writes s.reqs again in the setup of the next test.
* goroutine B reads it again in the callback from publishes in the tests.

So even though we've unsubscribed, there's no actual synchronisation around the reads and writes of s.reqs.

Reply:
(Demonstrated to @howbazaar to convince him.)